File: backend/app/models/ticket.py
from sqlalchemy import Column, Integer, String, Enum, DateTime, ForeignKey, Float, Text, Boolean
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from ..core.database import Base


class TicketPriority(str, enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class TicketStatus(str, enum.Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    CLOSED = "closed"
    ON_HOLD = "on_hold"


class Ticket(Base):
    __tablename__ = "tickets"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=False)
    priority = Column(Enum(TicketPriority), default=TicketPriority.MEDIUM, nullable=False)
    status = Column(Enum(TicketStatus), default=TicketStatus.OPEN, nullable=False)
    
    # Foreign keys
    client_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    assigned_agent_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    contract_id = Column(Integer, ForeignKey("contracts.id"), nullable=True)
    sla_id = Column(Integer, ForeignKey("slas.id"), nullable=True)
    
    # Pricing
    base_price = Column(Float, default=0.0)
    weight_multiplier = Column(Float, default=1.0)
    final_price = Column(Float, default=0.0)
    
    # SLA tracking
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    resolved_at = Column(DateTime, nullable=True)
    closed_at = Column(DateTime, nullable=True)
    due_date = Column(DateTime, nullable=True)
    sla_breached = Column(Boolean, default=False)
    
    # Relationships
    client = relationship("User", foreign_keys=[client_id], back_populates="tickets_created")
    assigned_agent = relationship("User", foreign_keys=[assigned_agent_id], back_populates="tickets_assigned")
    contract = relationship("Contract", back_populates="tickets")
    sla = relationship("SLA", back_populates="tickets")

File: frontend/src/types/ticket.ts
export enum TicketPriority {
  LOW = "low",
  MEDIUM = "medium",
  HIGH = "high",
  CRITICAL = "critical",
}

export enum TicketStatus {
  OPEN = "open",
  IN_PROGRESS = "in_progress",
  RESOLVED = "resolved",
  CLOSED = "closed",
  ON_HOLD = "on_hold",
}

export interface Ticket {
  id: number;
  title: string;
  description: string;
  priority: TicketPriority;
  status: TicketStatus;
  client_id: number;
  assigned_agent_id?: number;
  contract_id?: number;
  sla_id?: number;
  base_price: number;
  weight_multiplier: number;
  final_price: number;
  created_at: string;
  updated_at: string;
  resolved_at?: string;
  closed_at?: string;
  due_date?: string;
  sla_breached: boolean;
}

export interface TicketCreate {
  title: string;
  description: string;
  priority: TicketPriority;
  contract_id?: number;
}

export interface TicketUpdate {
  title?: string;
  description?: string;
  priority?: TicketPriority;
  status?: TicketStatus;
  assigned_agent_id?: number;
}
